service/history/historyEngine.go

```diff
@@ -696,6 +713,138 @@ func (e *historyEngineImpl) startWorkflowHelper(
 	}, nil
 }
 
+func shouldTerminateAndStart(
+	startRequest *h.StartWorkflowExecutionRequest,
+	state int,
+) bool {
+	return startRequest.StartRequest.GetWorkflowIdReusePolicy() == workflow.WorkflowIdReusePolicyTerminateIfRunning &&
+		(state == persistence.WorkflowStateRunning || state == persistence.WorkflowStateCreated)
+}
+
+// terminate running workflow then start a new run in one transaction
+func (e *historyEngineImpl) terminateAndStartWorkflow(
+	runningWFCtx workflowContext,
+	workflowExecution workflow.WorkflowExecution,
+	domainEntry *cache.DomainCacheEntry,
+	domainID string,
+	startRequest *h.StartWorkflowExecutionRequest,
+	signalWithStartRequest *h.SignalWithStartWorkflowExecutionRequest,
+) (*workflow.StartWorkflowExecutionResponse, error) {
+	runningMutableState := runningWFCtx.getMutableState()
+UpdateWorkflowLoop:
+	for attempt := 0; attempt < conditionalRetryCount; attempt++ {
+		if !runningMutableState.IsWorkflowExecutionRunning() {
+			return nil, ErrWorkflowCompleted
+		}
+
+		if err := execution.TerminateWorkflow(
+			runningMutableState,
+			runningMutableState.GetNextEventID(),
+			TerminateIfRunningReason,
+			getTerminateIfRunningDetails(workflowExecution.GetRunId()),
+			TerminateIfRunningIdentity,
+		); err != nil {
+			if err == ErrStaleState {
+				// Handler detected that cached workflow mutable could potentially be stale
+				// Reload workflow execution history
+				runningWFCtx.getContext().Clear()
+				if attempt != conditionalRetryCount-1 {
+					_, err = runningWFCtx.reloadMutableState()
+					if err != nil {
+						return nil, err
+					}
+				}
+				continue UpdateWorkflowLoop
+			}
+			return nil, err
+		}
+
+		// new mutable state
+		newMutableState, err := e.createMutableState(domainEntry, workflowExecution.GetRunId())
+		if err != nil {
+			return nil, err
+		}
+
+		if signalWithStartRequest != nil {
+			startRequest = getStartRequest(domainID, signalWithStartRequest.SignalWithStartRequest)
+		}
+
+		err = e.addStartEventsAndTasks(
+			newMutableState,
+			workflowExecution,
+			startRequest,
+			signalWithStartRequest,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		updateErr := runningWFCtx.getContext().UpdateWorkflowExecutionWithNewAsActive(
```

Comment:
LGTM. This update will call updateWorkflowExecutionEventReapply and reapply signals. Just double check it compatible with signal with start